### ISSUE TITLE
Q: Delay odd API

### ISSUE THREAD

**Issue by jrochkind** [ruby-concurrency/concurrent-ruby#513]
Concurrent::Delay has some odd API. 

Before it's been triggered at all, it's state is pending, and it's correspondingly not complete. OK. 

```
 delay = Concurrent::Delay.new {  sleep 20 ; "done" }
 delay.state # => "pending"
 delay.complete? # => false
```

Good enough. However, after it's triggered, both #state and #complete? become blocking, waiting on value. 

```
Thread.new {  delay.value }  # triggered it to start
delay.state # => won't return for 20 seconds, at which point it's :fulfilled
# OR
delay.complete? # => won't return for 20 seconds, at which point it's true
```

It seems to me the point of a Delay is that once it's created, I don't have to know whether it's been triggered yet or not. #value returns it's value either way (although waiting longer in some cases). 

But it's odd that #state blocks on completion only once it's been triggered.  This makes `state` and `complete` methods that should never be called -- if you don't want them to block on completion, they can't be called safely, because you don't know if the Delay has been triggered yet. If the Delay already was triggered (which you have no way to safely check, because those methods would be it, but they're not), then they will never return until state is :fulfilled or :rejected. 

I would think there should be a way to tell the difference between a Delay that hasn't been triggered yet at all, and one that has been triggered but isn't complete yet. But `pending` means that the Delay hasn't been triggered yet -- and there's no way to see state of a Delay that has been triggered but not yet complete, as methods that might tell you that block for completion instead. 

What do you think?

**Comment by jdantonio** [comment]
The API for Delay is definitely wonky. It started out as a simple class based on Clojure's `delay` function. But then we started adding more features. Then we decided to align it with `IVar`, `Future`, etc. And now it's not quire so simple. In 2.0 we plan to have a simpler API and fewer features. Anyway...

It seems to me that `#state` and `complete?` should not be blocking in the way you describe. I've taken a quick look at the code and I see that the root cause is the `Obligation` mixin. A fix would be easy, but I now wonder if `Future`, etc. behave similarly. `IVar` includes `Obligation` as well. I'll need to look into this further.
